### PR TITLE
[codegen] using TORCH_LIBRARY_FRAGMENT for some namespaces

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #88229

Summary:
Sometimes we want to extend an existing custom namespace library, instead of creating a new one,
but we don't have a namespace config right now, so we hardcode some custom libraries defined
in pytorch today, i.e. quantized and quantized_decomposed

Test Plan:
ci

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

**Comment by larryliu0820** [comment]
Can you fix the failing test?

**Comment by jerryzh168** [comment]
> Can you fix the failing test?

done, please take a look again

**Comment by ezyang** [comment]
IMO, this should be configurable on a per namespace basis. If custom namespace lives entitely in yaml it is reasonable to want it to be closed.

**Comment by larryliu0820** [comment]
> IMO, this should be configurable on a per namespace basis. If custom namespace lives entitely in yaml it is reasonable to want it to be closed.

I think one way to support this is to add a tag for a function. For example,

```
- func: quantized::add.out(Tensor a, ..., *, Tensor(a!) out) -> Tensor(a!)
  variants: function
  dispatch:
    CPU: quantized_add_out
  tags: fragment
```

This means that we are expecting a `quantized` library already exist in PyTorch (`quantized/library.cpp`) and just extending it by adding a new `add.out` operator.

**Comment by jerryzh168** [comment]
> > IMO, this should be configurable on a per namespace basis. If custom namespace lives entitely in yaml it is reasonable to want it to be closed.
> 
> I think one way to support this is to add a tag for a function. For example,
> 
> ```
> - func: quantized::add.out(Tensor a, ..., *, Tensor(a!) out) -> Tensor(a!)
>   variants: function
>   dispatch:
>     CPU: quantized_add_out
>   tags: fragment
> ```
> 
> This means that we are expecting a `quantized` library already exist in PyTorch (`quantized/library.cpp`) and just extending it by adding a new `add.out` operator.

should this config be library specific instead of operator specific?

File: torchgen/gen.py [hunk 1]
```diff
@@ -1623,8 +1623,11 @@ def get_native_function_schema_registrations(
     schema_selector: SelectiveBuilder,
 ) -> Tuple[List[str], str]:
     ns_native_functions: Dict[str, List[NativeFunction]] = defaultdict(list)
+    # a map from namespace to whether we should use TORCH_LIBRARY or TORCH_LIBRARY_FRAGMENT
+    ns_to_is_fragment: Dict[str, bool] = defaultdict(bool)
     for native_function in native_functions:
         ns_native_functions[native_function.namespace].append(native_function)
+        ns_to_is_fragment[native_function.namespace] |= "fragment" in native_function.tags
```
Comment (larryliu0820) [inline torchgen/gen.py:1630]: Add it here: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/tags.yaml

File: tools/test/test_codegen.py [hunk 2]
```diff
@@ -217,6 +217,11 @@ def setUp(self) -> None:
             loc=torchgen.model.Location(__file__, 1),
             valid_tags=set(),
         )
+        self.fragment_custom_native_function, _ = torchgen.model.NativeFunction.from_yaml(
+            {"func": "custom::func() -> bool"},
+            loc=torchgen.model.Location(__file__, 1),
+            valid_tags=set("fragment"),
+        )
```
Comment (larryliu0820) [inline tools/test/test_codegen.py:224]: Other than that can you add a test to parse `tags: fragment` from yaml file?

Comment (jerryzh168) [reply tools/test/test_codegen.py:224]: OK, sounds good

**Comment by ezyang** [comment]
The config should not live on an op. Should be somewhere library info.

File: torchgen/gen.py [hunk 3]
```diff
@@ -1640,8 +1640,16 @@ def get_native_function_schema_registrations(
         else:
             custom_namespace = namespace
             tab = "\t"
+            PREDEFINED_NAMESPACES = set(["quantized", "quantized_decomposed"])
```
Comment (ezyang) [inline torchgen/gen.py:1643]: You can use curly here to make a set literal

Comment (ezyang) [reply torchgen/gen.py:1643]: Preferably put this top level and better name eg FRAGMENT_NAMESPACES

**Review by ezyang — APPROVED** [review]
I did not review tests

File: torchgen/gen.py (hunk 3 above)
Comment (larryliu0820) [reply torchgen/gen.py:1643]: Also can we add it to https://github.com/pytorch/pytorch/blob/99c07735e457a2961f2319b4ba19f0d04eb47967/torchgen/model.py#L49

**Pull request closed without merging**